Adds a nonlinear preconditioned grain growth model

tonkmr opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
Adds nonlinear preconditioning to the grain growth model, to reduce the required mesh size. closes #21940.

## Design
It adds:
- New kernels
- New ics
- A new material class
- An action that automates it all
- tests
- All documentation

## Impact
Does not impact any existing models, but enables more efficient grain growth simulations

tonkmr committed:
Adds a nonlinear preconditioned grain growth model

It adds:
- New kernels
- New ics
- A new material class
- An action that automates it all
- tests
- All documentation

closes #21938 and #21940.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1569242/) on de9fe4d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24599/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24599/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 968eb46a3868f624d7930083ffdedd039b3d4fd5`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1569434/) on 3ca614f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24599/site)

This comment will be updated on new commits.

tonkmr commented:
@permcody I added a GrainTracker test for the new functionality and it is failing in some cases, but not all. It passes fine on my computer. It looks like it is failing because on some systems GrainTracker remaps to a different empty variable than on other systems. This isn't necessarily wrong, its just different. Do you have any recommendations? I could just use fewer order parameters so it only has one option.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1578365/) on 8394508 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>c4ce99</th>
      <th colspan=4>#24599 839450</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.88%</td>
      <td>86.09%</td>
      <td>+0.21%</td>
      <td>95.48%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12035</td>
      <td>12336</td>
      <td>+301</td>
      <td>317</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1979</td>
      <td>1994</td>
      <td>+15</td>
      <td>15</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24599/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24599/coverage/phase_field)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24599/coverage/xfem)

</details>



This comment will be updated on new commits.

tonkmr commented:
@dschwen a good place to start with this PR would be the new documentation page Nonlin_Precond_Grain_Growth.md. It summarizes the approach and all of the new objects that I have created to implement it.

tonkmr commented:
@dschwen Check SQA is failing on the Documentation test. It says `/AuxKernels/NonlinPrecondAux` is missing a markdown file, but it isn't. I have added that file. It seems to be looking for it in `framework` instead of `modules` but I'm not sure why.

GiudGiud commented on modules/phase_field/src/auxkernels/NonlinPrecondAux.C:
```diff
@@ -0,0 +1,34 @@
+
+#include "NonlinPrecondAux.h"
+
+registerMooseObject("MooseApp", NonlinPrecondAux);
```
this should be phase field app not MooseApp

this is why SQA is failing

GiudGiud replied:
(eg because it sees mooseApp, it looks in framework/doc instead of modules...doc

tonkmr commented:
@laagesen This is the grain growth nonlinear preconditioning model that I have been working on. Please review.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1609273/) on 3591928 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

dschwen commented on modules/phase_field/include/actions/GrainGrowthNonlinPrecondAction.h:
```diff
@@ -0,0 +1,40 @@
+
+// Forward declaration
+
+class GrainGrowthNonlinPrecondAction : public Action
```
This looks similar enough to `GrainGrowthAction` that some refactoring would be in order.

tonkmr replied:
I addressed this comment and refactored GrainGrowthAction and now this action inherits from it.

dschwen commented on modules/phase_field/include/kernels/ACGrGrPoly.h:
```diff
@@ -26,6 +26,8 @@ class ACGrGrPoly : public ACGrGrBase
   ACGrGrPoly(const InputParameters & parameters);
 
 protected:
+  virtual Real assign_this_op();
+  virtual std::vector<Real> assign_other_ops();
```
```suggestion
  virtual Real assignThisOP();
  virtual std::vector<Real> assignOtherOPs();
```

tonkmr replied:
@dschwen I have addressed this comment

dschwen commented on modules/phase_field/src/actions/PolycrystalColoringICAction.C:
```diff
@@ -19,20 +19,29 @@ InputParameters
   InputParameters params = Action::validParams();
-  params.addClassDescription("Random Voronoi tessellation polycrystal action");
+  params.addClassDescription(
+      "Action to create ICs for polycrystal variables defined using a Voronoi tesselation");
```
```suggestion
      "Action to create ICs for polycrystal variables defined using a Voronoi tessellation");
```

tonkmr replied:
@dschwen I have addressed this comment

dschwen commented on modules/phase_field/src/actions/PolycrystalColoringICAction.C:
```diff
@@ -19,20 +19,29 @@ InputParameters
+                        "Whether to use nonlinear preconditioning or the standard model");
+  params.addParam<Real>("bound_value",
+                        5.0,
+                        "ONLY FOR NONLINEAR PRECONDITIONING: Bound value used to keep variable "
```
Rather than the all cap comment how about putting both parameters into a `Nonlinear Preconditioning` parameter group?

tonkmr replied:
@dschwen I can't remember the syntax for parameter groups. Can you point me at an example?

GiudGiud replied:
params.addParamNamesToGroup("shift scale", "group name");

tonkmr replied:
I have addressed this comment

dschwen commented on modules/phase_field/src/actions/PolycrystalColoringICAction.C:
```diff
@@ -19,20 +19,29 @@ InputParameters
     _op_num(getParam<unsigned int>("op_num")),
-    _var_name_base(getParam<std::string>("var_name_base"))
+    _var_name_base(getParam<std::string>("var_name_base")),
+    _nonlin_precond(getParam<bool>("nonlinear_preconditioning"))
```
I'd love to see all instances of "nonlin precond" expanded to "nonlinear preconditioning"

tonkmr replied:
@dschwen do you mean just variable names like here or everywhere, including the file names?

tonkmr replied:
@dschwen @laagesen We talked in person about changing the name of this because "nonlinear preconditioning" might make users think this is related to preconditioning in PETSc. @laagesen suggested "change-of-variable" but I think that might be too generic. Here are some other options I have thought of:

- Linear Interface
- Linear Substitution
- Linear Variable Substitution

Which do you prefer?

dschwen replied:
The substitution is not linear, but it _linearizes_ the order parameters. `LinearizedInterface` would be my sugegstion.

tonkmr replied:
I like LinearizedInterface

laagesen replied:
I also like `LinearizedInterface`.

dschwen commented on modules/phase_field/src/auxkernels/NonlinPrecondAux.C:
```diff
@@ -0,0 +1,34 @@
+Real
+NonlinPrecondAux::computeValue()
+{
+  return (1 + std::tanh(_phi[_qp] / std::sqrt(2.0))) / 2.0;
```
```suggestion
  return (1.0 + std::tanh(_phi[_qp] / std::sqrt(2.0))) / 2.0;
```

tonkmr replied:
I have addressed this comment

dschwen commented on modules/phase_field/src/kernels/ACGrGrPolyNonlinPrecond.C:
```diff
@@ -0,0 +1,96 @@
+{
+  switch (type)
+  {
+    case Residual:
+    {
+      return ACGrGrPoly::computeDFDOP(type);
+    }
+
+    case Jacobian:
+    {
+      return ACGrGrPoly::computeDFDOP(type) * _dopdphi[_qp];
+    }
```
```suggestion
    case Residual:
      return ACGrGrPoly::computeDFDOP(type);

    case Jacobian:
      return ACGrGrPoly::computeDFDOP(type) * _dopdphi[_qp];
```

tonkmr replied:
I have addressed this comment

dschwen commented on modules/phase_field/src/utils/ExpressionBuilder.C:
```diff
@@ -256,6 +257,7 @@ UNARY_FUNC_IMPLEMENT(log10, LOG10)
 UNARY_FUNC_IMPLEMENT(exp, EXP)
 UNARY_FUNC_IMPLEMENT(sinh, SINH)
 UNARY_FUNC_IMPLEMENT(cosh, COSH)
+UNARY_FUNC_IMPLEMENT(tanh, TANH)
```
:+1:

tonkmr commented:
@GiudGiud @dschwen  Two internal checks are failing and I am not able to see the results. Can you tell me what the errors are?

GiudGiud commented:
Seems marmot does not build anymore
@dschwen

tonkmr commented:
I assume that there is another action in MARMOT that inherits from GrainGrowthAction, but I don't have access to MARMOT to fix it.

laagesen commented:
@GiudGiud I tried to look at the internal app tests and cannot see them, not sure if this might be related to the switch over to INL Github, I will contact you via Teams about this...

@tonkmr Did you lose access to Marmot due to the switch from Gitlab to Github, or did it lapse before that? Do you want to get access restored?

loganharbour commented:
@laagesen the failure is in marmot. If you're signed in via github.com on civet, you should be able to see the results now.

dschwen commented:
Yeah, it's my `SwellingGrainGrowthAction` :-/

dschwen commented:
I'm working on a Marmot PR to fix this

laagesen commented:
thanks @dschwen - with Logan's help I can see the internal tests now

dschwen commented on modules/phase_field/include/actions/GrainGrowthAction.h:
```diff
@@ -24,9 +24,8 @@ class GrainGrowthAction : public Action
   virtual void act();
 
 protected:
-  void addKernel(const std::string & kernel_type,
-                 const std::string & kernel_name,
-                 InputParameters params);
```
Why is this being removed?

dschwen replied:
Hmm, and while you're at it would you mind making this virtual. I'm puzzled how this ever worked without being virtual...

dschwen replied:
actually scratch that

dschwen commented on modules/phase_field/src/actions/GrainGrowthAction.C:
```diff
@@ -146,7 +124,7 @@ GrainGrowthAction::act()
         params.applyParameters(parameters());
 
-        addKernel(kernel_type, kernel_name, params);
+        _problem->addKernel(kernel_type, kernel_name, params);
```
undo please

dschwen replied:
hold off on that

dschwen commented on modules/phase_field/src/actions/GrainGrowthAction.C:
```diff
@@ -223,33 +245,25 @@ GrainGrowthAction::act()
   }
 }
-
-void
-GrainGrowthAction::addKernel(const std::string & kernel_type,
-                             const std::string & kernel_name,
-                             InputParameters params)
-{
-  _problem->addKernel(kernel_type, kernel_name, params);
-}
```
Undo this please, this is what breaks Marmot!

dschwen replied:
I can fix Marmot instead. This function isn't really needed.

dschwen commented:
Marmot PR is ready.

tonkmr commented:
Thanks! I still need to rename everything, but then it will be ready to go.

Mike

--------------
Michael Tonks
Associate Chair and Alumni Professor
Materials Science and Engineering
University of Florida

100D Rhines Hall, Gainesville, FL 32611
(352) 846-3779
***@***.***


On Jul 10, 2023, at 5:19 PM, Daniel Schwen ***@***.***> wrote:

[External Email]


Marmot PR is ready.

—
Reply to this email directly, view it on GitHub<https://github.com/idaholab/moose/pull/24599#issuecomment-1629746605>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABSNSRCQOESDV3UXLECIIYTXPRWWPANCNFSM6AAAAAAY3IY6QQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

tonkmr commented:
@GiudGiud you mentioned on Slack that since I have a documentation page doc/source/actions/GrainGrowthLinearizedInterfaceAction.md, I do not need to create doc/syntax/Modules/PhaseField/GrainGrowthLinearizedInterface/index.md. However, the Documentation tests are failing with the error: "Unable to locate a page that ends with the name 'syntax/Modules/PhaseField/GrainGrowthLinearizedInterface/index.md". Any idea on what might be causing this? Thanks!

GiudGiud commented:
Seems I was wrong. You need a syntax documentation page because you introduced a new syntax

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1706433/) on 32130c4 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 90.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 82.0%)
- Failed to generate `heat_conduction` coverage rate (required: 87.0%)
- Failed to generate `misc` coverage rate (required: 34.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `scalar_transport` coverage rate (required: 90.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)

tonkmr commented:
@dschwen @laagesen @GiudGiud I have addressed all of the comments, so this is ready to review again. I would like to wait and transition the LinearizedInterfaceMaterial away from ExpressionBuilder in a new PR.

GiudGiud commented:
I'll let @dschwen decide on the next steps here

dschwen commented on modules/phase_field/test/tests/actions/gold/grain_growth_out.e:
Looks like those are not used :-/

GiudGiud replied:
I see the test exists already,.

are these regolds? It looks like a new file on git

GiudGiud replied:
I get this diff with exodiff:
```
  ==============================================================
  NOTE: All node and element ids are reported as global ids.

Sideset Distribution Factors:
  --------- Time step 1, 0.0000000e+00 ~ 0.0000000e+00, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:
   bnds  rel diff:  0.0000000e+00 ~  1.0000000e+00 = 1.00000e+00 (node 1)
  --------- Time step 2, 8.0000000e+01 ~ 8.0000000e+01, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:
  --------- Time step 3, 1.6000000e+02 ~ 1.6000000e+02, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:
  --------- Time step 4, 2.4000000e+02 ~ 2.4000000e+02, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:
  --------- Time step 5, 3.2000000e+02 ~ 3.2000000e+02, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:
  --------- Time step 6, 4.0000000e+02 ~ 4.0000000e+02, rel diff:  0.00000e+00 ---------
Global variables:
Nodal variables:

exodiff: Files are different
```

tonkmr replied:
I changed the action so that the BndsCalcAux runs on initial, when it didn't previously. So, I had to re-gold the tests. The input files are the same, but the action is different, so the files had to change.

GiudGiud replied:
alright thanks for double checking

dschwen reviewed (dismissed):
Please remove unused files.
